src/library.js

```diff
@@ -1205,14 +1206,17 @@ LibraryManager.library = {
   // an invalid index into the FUNCTION_TABLE, so something has gone wrong.
   __cxa_end_catch__deps: ['__cxa_free_exception', '$EXCEPTIONS'],
   __cxa_end_catch: function() {
-    if (___cxa_end_catch.rethrown) {
-      ___cxa_end_catch.rethrown = false;
+    var ptr = EXCEPTIONS.caught.pop();
+    if (ptr && EXCEPTIONS.infos[EXCEPTIONS.deAdjust(ptr)].rethrown) {
```

Comment:
Things diverge here because rethrown exceptions need processing that is different from ones that aren't rethrown. In particular, and as to your second question, they should not be destroyed even if the reference count drops to zero (which it may, or may not, depending on whether there are additional references to the rethrown exception). The lifetime sequence of a re-thrown exception is as follows: entering handler increases refcount; then re-throw occurs in handler; leaving handler decrements refcount (possibly to zero - and that's where we must be careful not to destroy); then unwinding occurs and some new handler will be entered, increasing refcount (possibly from zero); then that handler will terminate, decrementing refcount again (possibly for the final time, and thus destroying).